### PR TITLE
Release 0.5.0

### PR DESCRIPTION
**Pull request by blootsvoets**
Changes since version 0.4.0:
- Directly submit data to Apache Kafka, instead of via Confluent's REST Proxy
- Moved from Travis to Github Actions
- Docker-compose configuration in Gradle
- Use radar-jersey for authorisation and exception handling
- Modularized configuration

### CODE REVIEW

File: gateway.yml [hunk 1]
```diff
@@ -1,10 +1,43 @@
...
+    #isJmxEnabled: true
+
+kafka:
+    # Number of Kafka brokers to keep in a pool for reuse in multiple requests.
+    # poolSize: 20
+    # Kafka producer settings. Read from https://kafka.apache.org/documentation/#producerconfigs.
+    producer:
+        bootstrap.servers: kafka-1:9092
+    # Kafka Admin Client settings. Read from https://kafka.apache.org/documentation/#adminclientconfigs.
+    # admin: {}
```
Comment (nivemaham) [inline gateway.yml]: is this a map? how do we specify multiple properties?

Comment (blootsvoets) [reply gateway.yml]: Indeed, the same as for `producer`

File: build.gradle [hunk 2]
```diff
@@ -16,22 +14,19 @@ ext {
...
+    radarSchemasVersion = '0.5.10.1'
+    jacksonVersion = '2.11.0'
+    slf4jVersion = '1.7.30'
     logbackVersion = '1.2.3'
     grizzlyVersion = '2.4.4'
-    jerseyVersion = '2.28'
     lzfseVersion = '0.1.0'
-    jaxbVersion = '2.3.0'
-    javaxActivationVersion = '1.1.1'
+    kafkaVersion = "2.5.0"
```
Comment (nivemaham) [inline build.gradle:26]: Is this backward compatible? would this still work with older versions of confluent/kafka?

File: src/main/kotlin/org/radarbase/gateway/inject/SchemaRetrieverFactory.kt [hunk 3]
```diff
@@ -1,17 +1,22 @@
...
+        @Context private val config: Config
+): Supplier<SchemaRetriever> {
     override fun get(): SchemaRetriever {
-        return SchemaRetriever(ServerConfig(config.schemaRegistryUrl), 30)
+        val server = when (val schemaRegistryUrl = config.kafka.serialization[SCHEMA_REGISTRY_URL_CONFIG]) {
+            is String -> ServerConfig(schemaRegistryUrl)
+            is List<*> -> ServerConfig(schemaRegistryUrl.first() as String)
+            else -> throw IllegalStateException("Configuration does not contain valid schema.registry.url")
+        }
+        return SchemaRetriever(server, 30, 300)
```
Comment (nivemaham) [inline src/main/kotlin/org/radarbase/gateway/inject/SchemaRetrieverFactory.kt]: shouldn't we use the latest version where you can send your own RestClient with authentication?

File: src/main/kotlin/org/radarbase/gateway/io/AvroProcessor.kt [hunk 4]
```diff
@@ -1,21 +1,57 @@
...
+        @Context private val auth: Auth,
+        @Context private val schemaRetriever: SchemaRetriever,
+        @Context schedulingService: SchedulingService
+): Closeable {
+    private val idMapping: ConcurrentMap<Pair<String, Int>, CachedValue<JsonToObjectMapping>> = ConcurrentHashMap()
+    private val schemaMapping: ConcurrentMap<Pair<String, String>, CachedValue<JsonToObjectMapping>> = ConcurrentHashMap()
+    private val cleanReference: SchedulingService.RepeatReference
+
+    init {
+        cleanReference = schedulingService.repeat(SCHEMA_CLEAN, SCHEMA_CLEAN) {
```
Comment (nivemaham) [inline src/main/kotlin/org/radarbase/gateway/io/AvroProcessor.kt:51]: what are we scheduling to clean here or how does this work? could you please add some documentation?

File: src/main/kotlin/org/radarbase/gateway/io/AvroProcessor.kt [hunk 5]
```diff
@@ -26,70 +62,313 @@ class AvroProcessor(@Context private val auth: Auth) {
...
 
-        records.forEach { record ->
-            if (isNullField(record["value"])) {
-                throw InvalidContentException("Missing value field in record")
-            }
-            val key = record["key"] ?: throw InvalidContentException("Missing key field in record")
-            processKey(key, auth)
+        return records.map { record ->
+            val key = record["key"] ?: throw HttpInvalidContentException("Missing key field in record")
+            val value = record["value"] ?: throw HttpInvalidContentException("Missing value field in record")
```
Comment (nivemaham) [inline src/main/kotlin/org/radarbase/gateway/io/AvroProcessor.kt:99]: I'm curious why do we remove the check for `field.isNull`

**Review by nivemaham — APPROVED** [review]
LGTM. I left some minor comments. is this tested on any of our test servers?

File: src/main/kotlin/org/radarbase/gateway/io/AvroProcessor.kt (hunk 5 above)
Comment (blootsvoets) [reply src/main/kotlin/org/radarbase/gateway/io/AvroProcessor.kt:99]: The value type is now being checked in `processValue`. It wasn't checked elsewhere in the previous implementation.

File: src/main/kotlin/org/radarbase/gateway/io/AvroProcessor.kt (hunk 4 above)
Comment (blootsvoets) [reply src/main/kotlin/org/radarbase/gateway/io/AvroProcessor.kt:51]: added a comment, it's meant to regularly free up memory of stale caches.

File: build.gradle (hunk 2 above)
Comment (blootsvoets) [reply build.gradle:26]: I've run the integration tests for 4.1.4 up to 5.5.0, they all succeed.

File: src/main/kotlin/org/radarbase/gateway/inject/SchemaRetrieverFactory.kt (hunk 3 above)
Comment (blootsvoets) [reply src/main/kotlin/org/radarbase/gateway/inject/SchemaRetrieverFactory.kt]: Done, see https://github.com/RADAR-base/RADAR-Gateway/blob/7a71f39c1e1af0c60734f6f3c96ee8c7bfcbbcda/src/main/kotlin/org/radarbase/gateway/inject/SchemaRetrieverFactory.kt#L26 and below.

**Comment by blootsvoets** [comment]
This is running on our test server. Note that for now, a radar server ALSO needs the rest-proxy to be present since the radar-backend initialisation checks depend on it.

**Pull request merged**